src/NodeApi/JSValueScope.cs

```diff
@@ -174,35 +245,49 @@ public sealed class JSValueScope : IDisposable
                 // Module scopes may be created without a parent scope (for AOT modules).
                 if (scopeType != JSValueScopeType.Module)
                 {
-                    throw new InvalidOperationException("Parent scope not found.");
+                    throw new InvalidOperationException(
+                        $"A {scopeType} scope cannot be created without a parent scope.");
                 }
 
                 // AOT module scopes are constructed with an env parameter
                 // but without a pre-initialized runtime.
                 _env = env.IsNull ? throw new ArgumentNullException(nameof(env)) : env;
+                ThreadId = Environment.CurrentManagedThreadId;
                 Runtime = runtime ?? new NodejsRuntime();
             }
+            else if (_parentScope.IsDisposed)
+            {
+                // This should never happen because disposing a scope removes it from
+                // s_currentScope (which is used to initialize _parentScope above).
+                throw new InvalidOperationException("Parent scope is disposed.");
+            }
+            else if (scopeType == JSValueScopeType.Callback &&
+                _parentScope.ScopeType != JSValueScopeType.Callback &&
+                _parentScope.ScopeType != JSValueScopeType.Module &&
+                _parentScope.ScopeType != JSValueScopeType.Root &&
+                _parentScope.ScopeType != JSValueScopeType.NoContext)
+            {
+                throw new InvalidOperationException(
```

Comment:
>InvalidOperationException [](http://example.com/codeflow?start=26&length=25)

The condition above also checks for the NoContext scope type.

Reply:
Yes, but that scope type is not meant to be used by user code, so I didn't think it was worth mentioning in the exception message. (Root scope may be used in embedding scenarios and Module scope may be used in custom module initializers that don't use the generator, but I don't think NoContext scope is applicable anywhere outside the native host.)